subsys/bluetooth/services/bas/bas.c

```diff
@@ -0,0 +1,145 @@
+}
+
+static void blvl_status_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	ARG_UNUSED(attr);
+
+	bool notif_enabled = (value == BT_GATT_CCC_NOTIFY);
+	bool ind_enabled = (value == BT_GATT_CCC_INDICATE);
+
+	LOG_INF("BAS Notifications %s", notif_enabled ? "enabled" : "disabled");
+	LOG_INF("BAS Indications %s", ind_enabled ? "enabled" : "disabled");
+}
+
+static ssize_t read_blvl(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			 uint16_t len, uint16_t offset)
+{
+	uint8_t lvl8 = battery_level;
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &lvl8, sizeof(lvl8));
+}
+
+static void indicate_cb(struct bt_conn *conn, struct bt_gatt_indicate_params *params, uint8_t err)
+{
+	if (err != 0) {
+		LOG_INF("Indication failed with error %d\n", err);
+	} else {
+		LOG_INF("Indication sent successfully\n");
+	}
+}
+
+/* Constant values from the Assigned Numbers specification:
+ * https://www.bluetooth.com/wp-content/uploads/Files/Specification/Assigned_Numbers.pdf?id=89
+ */
+static const struct bt_gatt_cpf level_cpf = {
+	.format = 0x04, /* uint8 */
+	.exponent = 0x0,
+	.unit = 0x27AD,        /* Percentage */
+	.name_space = 0x01,    /* Bluetooth SIG */
+	.description = 0x0106, /* "main" */
+};
+
+BT_GATT_SERVICE_DEFINE(bas, BT_GATT_PRIMARY_SERVICE(BT_UUID_BAS),
+		       BT_GATT_CHARACTERISTIC(BT_UUID_BAS_BATTERY_LEVEL,
+					      BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY,
+					      BT_GATT_PERM_READ, read_blvl, NULL, NULL),
+		       BT_GATT_CCC(blvl_ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+		       BT_GATT_CPF(&level_cpf),
+		       BT_GATT_CHARACTERISTIC(BT_UUID_BAS_BATTERY_LEVEL_STATUS,
+					      BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY,
+					      BT_GATT_PERM_READ, read_blvl_status, NULL, NULL),
+		       BT_GATT_CCC(blvl_status_ccc_cfg_changed,
+				   BT_GATT_PERM_READ | BT_GATT_PERM_WRITE));
+
+static int bas_init(void)
+{
+	bt_bas_bls_init();
+#if defined(CONFIG_BT_BAS_BLS_IDENTIFIER_PRESENT)
+	bt_bas_bls_set_identifier(level_cpf.description);
+#endif
+	return 0;
+}
+
+uint8_t bt_bas_get_battery_level(void)
+{
+	return battery_level;
+}
+
+int bt_bas_set_battery_level(uint8_t level)
+{
+	int rc;
+
+	if (level > 100U) {
+		return -EINVAL;
+	}
+
+	battery_level = level;
+
+	rc = bt_gatt_notify(NULL, &bas.attrs[1], &level, sizeof(level));
+
+#if defined(CONFIG_BT_BAS_BLS_BATTERY_LEVEL_PRESENT)
+	bt_bas_bls_set_battery_level(level);
+#endif
+	return rc == -ENOTCONN ? 0 : rc;
+}
+
+void update_battery_level_status(struct bt_bas_bls *level_status)
+{
+	int err;
+
+	/* Notify/Indicate all connections */
+	ind_params.attr = &bas.attrs[BT_BAS_IDX_BATT_LVL_STATUS_CHAR_VAL];
+	ind_params.data = level_status;
+	ind_params.len = sizeof(struct bt_bas_bls);
+	ind_params.func = indicate_cb;
+	err = bt_gatt_indicate(NULL, &ind_params);
+	if (err) {
+		LOG_DBG("Failed to send ntf/ind to all connections (err %d)\n", err);
+	}
+}
```

Comment:
Should be guarded by `CONFIG_BT_BAS_BLS`

I assume it has to exist here rather in the `bas_bls.c` file due to `bas.attrs`? 

Might be easier to place this in `bas_bls.c` and make `bas.attrs` accesible from `bap_bls.c` instead

Reply:
Since the `bls` feature is not larger than it is, we could also keep it all in a single file, but I do kind of like that it's placed in a new file given that it's optional.